### ISSUE TITLE
Add a macro that transforms the conditions into portable conditions

### ISSUE THREAD

**Issue by fstamour** [phoe/external-symbol-not-found#3]
I already did a small experiment [here](https://github.com/fstamour/breeze/blob/fcaeb6dd5ec124ed5d57a06dbfa6fbada75ac2e0/scratch-files/external-symbol-not-found.lisp).

**Comment by phoe** [comment]
Note that `handler-case` is unusable here because it destroys the original stack trace - `handler-bind` will be necessary.

**Comment by fstamour** [comment]
[Updated version of the experiment](https://github.com/fstamour/breeze/blob/8379b90d51f310e0fd15575aba4e89759a7868df/scratch-files/external-symbol-not-found.lisp)

**Comment by phoe** [comment]
```lisp
(define-condition external-symbol-not-found (simple-error)
  ((package
    :reader external-symbol-not-found-package
    :initarg :package
    :documentation "The symbol's package.")
   (symbol
    :reader external-symbol-not-found-symbol
    :initarg :symbol
    :documentation "The name of symbol if it wasn't found. The symbol
    if it's not exported. This sould be 2 different conditions.")
   (condition
    :reader external-symbol-not-found-condition
    :initarg :condition
    :documentation "The implementation condition.")))

(handler-bind
    ((error
       #'(lambda (e)
	   (when (esnf:external-symbol-not-found-p e)
	     (let ((new-condition
	      (make-condition 'external-symbol-not-found
			      :package (esnf:external-symbol-not-found-package e)
			      :symbol (esnf:external-symbol-not-found-symbol-name e)
			      :condition e
			      :format-control (simple-condition-format-control e)
			      :format-arguments (simple-condition-format-arguments e))))
	       (error new-condition))))))
  (progn
    (unless (find-package 'temp)
      (make-package 'temp :use nil))
    (read-from-string "temp:symbol")))
```

You make the false assumption that `e` is a `simple-error` and it is valid to call `simple-condition-format-{control,arguments}` on it.

Also, `external-symbol-not-found` doesn't need to inherit from `simple-error` - simply write a custom `:report` for it that makes use of `package`, `symbol`, and `condition`.

Also, as you noted - we should differentiate between situations where a symbol is not found in the package altogether and where a symbol is found but is not exported.

**Comment by fstamour** [comment]
> You make the false assumption that e is a simple-error and it is valid to call simple-condition-format-{control,arguments} on it.

Yes, I did that conciously, it works on my sbcl :tm: . More seriously: that was just an experiment, I'll do the right thing :tm: later :tm:

> Also, external-symbol-not-found doesn't need to inherit from simple-error - simply write a custom :report for it that makes use of package, symbol, and condition.

- [ ] Inherit from `error`
- [ ] Make 2 conditions:
  - [ ] `symbol-not-exported`
  - [ ] `symbol-not-found`
- [ ] Make a macro to wrap forms with the `handler-bind`
- [ ] Tests for the macro

relatedly, are you fine with adding Parachute for the tests?

**Comment by phoe** [comment]
> relatedly, are you fine with adding Parachute for the tests?

Yep.

**Comment by fstamour** [comment]
There's only one thing that bothers me: what if I want to have "portable conditions" for other typed of error. This project's name wouldn't match.

I can implement the things we talked so far (I'll have to do it one way or another), but I think another system/package would be better if there are other conditions that I want to handle in similar manners.

**Comment by phoe** [comment]
No idea, to be honest. What kinds of conditions do you imagine?

The main issue with this approach is that these "portable conditions" are only possible to get within some dynamic scope, you need to "opt in" for them, so to speak, by establishing a handler that can recognize an implementation-dependent condition and then signal a portable one. That might be OK for some use cases but a no-go for others now that I think of it.

**Comment by fstamour** [comment]
> What kinds of conditions do you imagine?

On the top of my head, there's ` sb-pcl:class-not-found-error`. There's probably something when there's a bad type specifier, or when a slot doesn't exists on an object, etc. That kind of errors.

> but a no-go for others now that I think of it.

Good point, a macro that supports all the different conditions is probably not that useful, unless you can specify which conditions you want it to handle. It's feasible, I'll surely do some more experiments to figure out the best syntax.

**Comment by phoe** [comment]
> unless you can specify which conditions you want it to handle

I assume such a mechanism could be programmable, in a way - supply it with a predicate that answers whether a new condition object should be constructed, and a way of constructing a new condition object based on the old one, and you're good to go.

Your code already has it - `esnf:external-symbol-not-found-p` is the predicate and the whole `(let ((new-condition ...)) ...)` is a "converter".
